### ISSUE TITLE
FileSystemException: Creation of temporary directory failed, path = '.../flutter_tools.' (OS Error: No space left on device, errno = 28)

### ISSUE THREAD

**Issue by andrewkolos** [flutter/flutter#153777]
On 3.24.0 as of 8/20: 67 reports by 20 unique clients.

Possibly shares a root cause with https://github.com/flutter/flutter/issues/142537.

Flutter CLI commands (at least from first 10 reports): `flutter run`, `flutter test integration_test`

I'm fairly certain the tool has constructs to handle out-of-storage errors in its `ErrorHandling*` file system abstractions, but something might be leaking here.

```
FileSystemException: Creation of temporary directory failed, path = '/var/folders/7j/0l_gjs0d7vx87_j12jr1zk3m0000gq/T/flutter_tools.' (OS Error: No space left on device, errno = 28)
at _Directory.createTempSync(directory_impl.dart:170)
at ForwardingDirectory.createTempSync(forwarding_directory.dart:28)
at LocalFileSystem.systemTempDirectory(file_system.dart:236)
at ForwardingFileSystem.systemTempDirectory(forwarding_file_system.dart:33)
at IOSCoreDeviceControl._listCoreDevices(core_devices.dart:65)
at IOSCoreDeviceControl.getCoreDevices(core_devices.dart:122)
at XCDevice.getAvailableIOSDevices(xcdevice.dart:487)
at <asynchronous gap>(async)
at PollingDeviceDiscovery._populateDevices(device.dart:547)
at <asynchronous gap>(async)
at Future.wait.<anonymous closure>(future.dart:524)
at <asynchronous gap>(async)
at DeviceManager.getAllDevices(device.dart:198)
at <asynchronous gap>(async)
at DeviceBasedDevelopmentArtifacts.requiredArtifacts(flutter_command.dart:1908)
at <asynchronous gap>(async)
at FlutterCommand.verifyThenRunCommand(flutter_command.dart:1692)
at <asynchronous gap>(async)
at FlutterCommand.run.<anonymous closure>(flutter_command.dart:1394)
at <asynchronous gap>(async)
at AppContext.run.<anonymous closure>(context.dart:153)
at <asynchronous gap>(async)
at CommandRunner.runCommand(command_runner.dart:212)
at <asynchronous gap>(async)
at FlutterCommandRunner.runCommand.<anonymous closure>(flutter_command_runner.dart:372)
at <asynchronous gap>(async)
at AppContext.run.<anonymous closure>(context.dart:153)
at <asynchronous gap>(async)
at FlutterCommandRunner.runCommand(flutter_command_runner.dart:308)
at <asynchronous gap>(async)
at run.<anonymous closure>.<anonymous closure>(runner.dart:130)
at <asynchronous gap>(async)
at AppContext.run.<anonymous closure>(context.dart:153)
at <asynchronous gap>(async)
at main(executable.dart:93)
at <asynchronous gap>(async)
```

**Comment by andrewkolos** [comment]
Oh okay. I figured out what is happening here.

The filesystem called upon in `IOSCoreDeviceControl._listCoreDevices` should be a `ErrorHandlingFileSystem`. `ErrorHandlingFileSystem` has an override for `systemTempDirectory` that wraps the underlying `Directory` instance into a `ErrorHandlingDirectory`, which provides us the error handling we expect (which should catch `OS Error: No space left on device, errno = 28` and throw a `ToolExit`—causing the tool to exit gracefully instead of crashing).

https://github.com/flutter/flutter/blob/30835d7dac4c2df11c5a8d49dc5a9c2e3cfc5dbe/packages/flutter_tools/lib/src/base/error_handling_io.dart#L126-L128

Notice `delegate` here. What I think is happening is that `delegate` is an instance of [`LocalFileSystem`](https://github.com/flutter/flutter/blob/45c8881eb289e3b68e890cefa0acd13dbf800147/packages/flutter_tools/lib/src/base/file_system.dart#L187) (not to be confused with the `LocalFileSystem` provided by package:file). [`LocalFileSystem`'s override of `systemTempDirectory`](https://github.com/flutter/flutter/blob/45c8881eb289e3b68e890cefa0acd13dbf800147/packages/flutter_tools/lib/src/base/file_system.dart#L229) will actually create and return a new directory within the actual system temp directory. This is the source of the `at _Directory.createTempSync(directory_impl.dart:170)` at the top of the stack.

TL;DR: merely referencing `FileSystem.systemTempDirectory` can result in a create operation, so we need to wrap this in the error-handling logic within `ErrorHandlingFileSystem.systemTempDirectory`.
